src/game.cpp

```diff
@@ -9629,6 +9575,125 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
+        debugmsg("Tried to handle_liquid a non-liquid!");
+        // "canceled by the user" because we *can* not handle it.
+        return transfer_ok;
+    }
+
+    const auto create_activity = [&]() {
+        if( source_veh != nullptr ) {
+            u.assign_activity( activity_id( "ACT_FILL_LIQUID" ) );
+            serialize_liquid_source( u.activity, *source_veh, liquid.typeId() );
+            return true;
+        } else if( source_pos != nullptr ) {
+            u.assign_activity( activity_id( "ACT_FILL_LIQUID" ) );
+            serialize_liquid_source( u.activity, *source_pos, liquid );
+            return true;
+        } else if( source_mon != nullptr ) {
+            u.assign_activity( activity_id( "ACT_FILL_LIQUID" ) );
+            serialize_liquid_source( u.activity, *source_mon, liquid );
+            return true;
+        } else {
+            return false;
+        }
+    };
+
+    switch( target.dest_opt ) {
+    case LD_CONSUME:
+        u.consume_item( liquid );
+        transfer_ok = true;
+        break;
+    case LD_ITEM: {
+        item *const cont = target.item_loc.get_item();
+        const int item_index = u.get_item_position( cont );
+        // Currently activities can only store item position in the players inventory,
+        // not on ground or similar. TODO: implement storing arbitrary container locations.
+        if( item_index != INT_MIN && create_activity() ) {
+            serialize_liquid_target( u.activity, item_index );
+        } else if( u.pour_into( *cont, liquid ) ) {
+            if( cont->needs_processing() ) {
+                // Polymorphism fail, have to introspect into the type to set the target container as active.
+                switch( target.item_loc.where() ) {
+                case item_location::type::map:
+                    m.make_active( target.item_loc );
+                    break;
+                case item_location::type::vehicle:
+                    m.veh_at( target.item_loc.position() )->vehicle().make_active( target.item_loc );
+                    break;
+                case item_location::type::character:
+                case item_location::type::invalid:
+                    break;
+                }
+            }
+            u.mod_moves( -100 );
+        }
+        transfer_ok = true;
+        break;
+    }
+    case LD_VEH:
+        if( target.veh == nullptr ) {
+            break;
+        }
+        if( create_activity() ) {
+            serialize_liquid_target( u.activity, *target.veh );
+        } else if( u.pour_into( *target.veh, liquid ) ) {
+            u.mod_moves( -1000 ); // consistent with veh_interact::do_refill activity
+        }
+        transfer_ok = true;
+        break;
+    case LD_KEG:
+    case LD_GROUND:
+        add_msg( "filling a %s", target.dest_opt == LD_KEG ? "keg" : "ground" );
+        if( create_activity() ) {
+            add_msg( "created a an activity" );
```

Comment:
If this is intended to remain, will need a new context as it is a bit abrupt and strange.